### DISCUSSION TITLE
[Performance] Full build takes more time since 2024-09

### DISCUSSION THREAD

**Discussion by fedejeanne** [eclipse-jdt/eclipse.jdt.core#3305] [Q&A] [answered]
Hi,
I've been looking into some performance issues that started with **2024-09**, specifically the time it takes to perform a **full build**. Interestingly enough, the time it takes to do a _clean all_ also increased in 2024-09 but the biggest increase happened in the build phase.

- In **2024-06** it takes ~12 minutes to do a full clean and build.
- In **2024-09** it takes ~30 minutes to do a full build (sampled **after** the _clean all_)

[samples.zip](https://github.com/user-attachments/files/17728799/samples.zip)

The part that caught my eye was the increase in amount of calls (_Hits_) and therefore the _Self Time (CPU)_ for the method(s) `org.eclipse.jdt.internal.compiler.lookup.TypeBinding.findSuperTypeOriginatingFrom(...)` (there are 2 flavors of this method).

**2024-06**
- 431 hits
- ~20s spent in these methods
![image](https://github.com/user-attachments/assets/2b876849-c0d4-4546-82b4-86625cacaf11)

**2024-09**
- 6972 hits
- ~9m spent in these methods
![image](https://github.com/user-attachments/assets/b809a235-e0da-44d0-bf7e-26a816497bfd)

## Question 1
- Why are there suddenly so many more calls to these methods ?

## Question 2
- Would it be OK to simply build a cache to speed up some subsequent repeated calls or would this introduce any undesired effects (_e.g._ inconsistencies)?

I ask because I noticed that many of the calls are "repeated" _i.e._ the input parameters are the same so a cache would be a trivial optimization.

**Comment by iloveeclipse** [comment]
@fedejeanne : could you please share the project that reproduces the performance degradation, or provide something where one can see the regression?

If you can't share the code, could you please bisect the regression down to the commit? Or could you please check whether this is also reproducible on Linux (I see Windows specific platform code in the profiler).

@jukzi : since you work on performance and especially on Windows, did you noticed any degradation like described here?

**Reply by fedejeanne** [reply]
I can't share the code since it's our tool, but I can tell you it has ~2000 projects (plugins, fragments and features).

I can't easily test in Linux, but I can tell you the the code is inside folders that are excluded from the virus scanner, so that shouldn't be an issue. I will try to provide more details regarding the Windows side and come back to Linux if necessary.

**Reply by jukzi** [reply]
> o that shouldn't be an issue

Unfortunately there may be a difference between "should" not and "is" not. The virus scanner may be informed a file is accessed and then the scanner may take a while to decide what todo. And some do even inject code in the running java application :-(

**Comment by jukzi** [comment]
Thanks for notifying me. Its interesting for us. I am not aware of such regression(s).
Please split it into two separate issues:
1. clean
2. findSuperTypeOriginatingFrom

- Please add also timings with current i-build (2024-12). 
- When sampling with VisualVm make sure to not have `java.**` excluded, to see the filesystem hotspots that are inside the JDK. I use this config:
![image](https://github.com/user-attachments/assets/703dc80a-a837-416c-b38c-87dc8fe363fc)
- Make sure to add the *.nps result for download.
- Give a public available reproducer.

**Reply by fedejeanne** [reply]
I will look into it an try to provide the samples/profiling.

What would you recommend for me to profile for both scenarios (clean + build)?

![image](https://github.com/user-attachments/assets/67f9fbea-d39b-4822-aa11-81f9f6841d77)

In any case, I am creating samples for both scenarios using 2024-06 and the latest I-BUILD. I can include those too if you want.

**Reply by jukzi** [reply]
> What would you recommend for me to profile for both scenarios (clean + build)?

i almost never profile but always sample:
![image](https://github.com/user-attachments/assets/a6b3482b-553b-4c9d-9b40-27dfe27d4b87)
Profile costs too much overhead, the timings are inaccurate, instrumenting the code takes too long. Only makes sense if one needs a call count for a certain class. But normally one does not know in advance what one should search for.

**Reply by fedejeanne** [reply]
Sorry, my wires got crossed and I read "profile" 😅 

Good, I'm sampling with this configuration (I increased to a frequency of **20ms** to avoid degrading the performance of the application):

![image](https://github.com/user-attachments/assets/8cd7f80f-8714-4280-837e-dfeba07d6b57)

I got these results:
- 2024-06: 2nd Clean all (after full build) = ~40 s
- 2024-06: Build all (clean workspace): = ~10 m
- I-BUILD 20241118-1800: 2nd Clean all (after full build) = ~36 s
- I-BUILD 20241118-1800: Build all (clean workspace): = ~33 m

Files: [Samples.zip](https://github.com/user-attachments/files/17814083/Samples.zip)


Notice that I sampled the _clean_ operation **after** the full build in order to take advantage of the "2nd access" that you mentioned [here](https://github.com/eclipse-jdt/eclipse.jdt.core/discussions/3305#discussioncomment-11290405). I don't think this plays a role because the code is always inside a folder that is excluded from the virus scanner but I did it anyway.

**Reply by iloveeclipse** [reply, marked as answer]
I've created https://github.com/eclipse-jdt/eclipse.jdt.core/issues/3327.
This looks really bad. Let close this discussion and continue with the bug.

**Comment by jukzi** [comment]
And please also note, that file access /(during clean) heavyly depends on the filesystem, cache and virus scanner(s), so that typically the first access takes far longer then a second access. So be sure to measure the same scenario (either a 2nd round or a first round after creating all the files from scratch)

**Comment by iloveeclipse** [comment]
Ideally you could test on Linux too, because your will have no antivirus scanner overhead and we could also sort out whether this is Windows specific IO issue or JDT core regression.

**Comment by jukzi** [comment]
The allSuperPairsWithCommonGenericType shown in sampling screenshots has been modified in that time by cc0e66ea57dbcef4054c360d84d2a0ecf8e9b5d4

**Reply by iloveeclipse** [reply]
@stephan-herrmann : FYI

**Comment by iloveeclipse** [comment]
Let continue on https://github.com/eclipse-jdt/eclipse.jdt.core/issues/3327.
